sdk/src/Core/Plugins/EC2Plugin.cs

```diff
@@ -60,36 +65,81 @@ public string ServiceName
         public bool TryGetRuntimeContext(out IDictionary<string, object> context)
         {
             context = null;
-
             var dict = new Dictionary<string, object>();
-            if (EC2InstanceMetadata.InstanceId != null)
+            const string metadata_base_url = "http://169.254.169.254/latest/";
+
+            try
             {
-                dict.Add("instance_id", EC2InstanceMetadata.InstanceId);
-            }
+                // try IMDSv2 endpoint for metadata
+                // get the token
+                Dictionary<string, string> header = new Dictionary<string, string>(1);
+                header.Add("X-aws-ec2-metadata-token-ttl-seconds", "60");
+                string token = DoRequest(metadata_base_url + "api/token", HttpMethod.Put, header).Result;
 
-            if (EC2InstanceMetadata.AvailabilityZone != null)
+                header = new Dictionary<string, string>(1);
+                header.Add("X-aws-ec2-metadata-token", token);
+                
+                // get the metadata
+                string resp = DoRequest(metadata_base_url + "dynamic/instance-identity/document", HttpMethod.Get, header).Result;
+                dict = ParseMetadata(resp);
+            }
+            catch (Exception e)
             {
-                dict.Add("availability_zone", EC2InstanceMetadata.AvailabilityZone);
+                _logger.Error(e, "Exception occurred while getting EC2 metadata from IMDSv2");
+                _logger.DebugFormat("Unable to get metadata from IMDSv2 endpoint. Falling back to IMDSv1 endpoint");
+
+                // try IMDSv1 endpoint for metadata
```

Comment:
I had a similar comment in the node PR - since there's no difference in the document endpoint for v2 and v1 I think we should just have error handling around the token request. The can continue with the same logic, including the token if it's available or otherwise try without it.

Reply:
Here's my version in Java which takes that approach, the flow is just

- get token
- get document (include token if we got one)
- parse document

Don't need to treat v1 as a different endpoint

Reply:
Oh yes. Good call out. The node PR discussion slipped out of my mind while writing this.